Sources/ComposableArchitecture/ViewStore.swift

```diff
@@ -433,47 +501,55 @@ private struct HashableWrapper<Value>: Hashable {
     ///   - animation: The animation to perform when the action is sent.
     ///   - predicate: A predicate on `State` that determines for how long this method should
     ///     suspend.
+    ///   - cancellationAction: An action that will be fired when the suspend task is cancelled.
     public func send(
       _ action: Action,
       animation: Animation?,
-      while predicate: @escaping (State) -> Bool
+      while predicate: @escaping (State) -> Bool,
+      onCancel cancellationAction: Action? = nil
     ) async {
       withAnimation(animation) { self.send(action) }
-      await self.suspend(while: predicate)
+      await self.suspend(while: predicate, onCancel: {
+        if let cancellationAction = cancellationAction {
+          self.send(cancellationAction)
+        }
+      })
     }
 
     /// Suspends while a predicate on state is `true`.
     ///
-    /// - Parameter predicate: A predicate on `State` that determines for how long this method
+    /// - Parameters:
+    ///   - predicate: A predicate on `State` that determines for how long this method
     ///   should suspend.
-    public func suspend(while predicate: @escaping (State) -> Bool) async {
-      if #available(iOS 15, macOS 12, tvOS 15, watchOS 8, *) {
-        _ = await self.publisher
-          .values
-          .first(where: { !predicate($0) })
-      } else {
-        let cancellable = Box<AnyCancellable?>(wrappedValue: nil)
-        try? await withTaskCancellationHandler(
-          handler: { cancellable.wrappedValue?.cancel() },
-          operation: {
-            try Task.checkCancellation()
-            try await withUnsafeThrowingContinuation {
-              (continuation: UnsafeContinuation<Void, Error>) in
-              guard !Task.isCancelled else {
-                continuation.resume(throwing: CancellationError())
-                return
-              }
-              cancellable.wrappedValue = self.publisher
-                .filter { !predicate($0) }
-                .prefix(1)
-                .sink { _ in
-                  continuation.resume()
-                  _ = cancellable
-                }
+    ///   - cancellationHandler: A closure that will be called if the suspend task is cancelled.
+    public func suspend(
+      while predicate: @escaping (State) -> Bool,
+      onCancel cancellationHandler: (() -> Void)? = nil
+    ) async {
+      let cancellable = Box<AnyCancellable?>(wrappedValue: nil)
+      try? await withTaskCancellationHandler(
+        handler: {
+          cancellable.wrappedValue?.cancel()
+          cancellationHandler?()
```

Comment:
Swift is warning that `cancellationHandler` isn't `@Sendable` - the same warning is reported for `cancellable` which is fixed by #987 - one possible solution here would be - following the merging of #987, to move the cancellation handler into `Box` so only that is being sent across and calling it that way.

Reply:
With the above fix in mind, I've refactored this so the Box holds on to the cancellation handler: https://github.com/pointfreeco/swift-composable-architecture/pull/1001/commits/8b16883df8b90350ce28d1b49e53938cfc75cc95